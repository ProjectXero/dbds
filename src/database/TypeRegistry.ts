import { factory, TypeNode } from 'typescript'

export const UNKNOWN = factory.createTypeReferenceNode('unknown')

export const defaultTypeMap = Object.freeze({
  bpchar: 'string',
  char: 'string',
  varchar: 'string',
  text: 'string',
  citext: 'string',
  uuid: 'string',
  bytea: 'string',
  inet: 'string',
  time: 'string',
  timetz: 'string',
  interval: 'string',
  name: 'string',
  int2: 'number',
  int4: 'number',
  int8: 'number',
  float4: 'number',
  float8: 'number',
  numeric: 'number',
  money: 'number',
  oid: 'number',
  bool: 'boolean',
  json: 'unknown',
  jsonb: 'unknown',
  date: 'Date',
  timestamp: 'Date',
  timestamptz: 'Date',
})

export default class TypeRegistry {
  public readonly typeMap: Record<string, string>

  constructor() {
    this.typeMap = { ...defaultTypeMap }
  }

  public isKnownType(typeName: string): boolean {
    return Object.prototype.hasOwnProperty.call(this.typeMap, typeName)
  }

  public getTypeNode(typeName: string): TypeNode {
    if (this.isKnownType(typeName)) {
      return factory.createTypeReferenceNode(this.typeMap[typeName])
    }

    console.warn(`Unknown type detected: '${typeName}'. This is probably a bug.`)
    this.registerType(typeName, 'unknown')

    return UNKNOWN
  }

  public registerType(typeName: string, target: string): void {
    if (this.isKnownType(typeName)) {
      console.warn(`Re-registering known type '${typeName}': ${this.typeMap[typeName]} => ${target}`)
    }

    this.typeMap[typeName] = target;
  }
}
