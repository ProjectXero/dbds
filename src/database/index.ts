export { default as SchemaInfo } from './SchemaInfo'
export * from './SchemaInfo'
export { default as TypeMapper } from './TypeMapper'
