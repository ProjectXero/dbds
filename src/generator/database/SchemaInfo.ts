import { DatabasePoolType, sql, TaggedTemplateLiteralInvocationType } from 'slonik'

export interface EnumInfo {
  name: string
  values: string[]
}

export interface ColumnInfo {
  name: string
  nullable: boolean
  hasDefault: boolean
  order: number
  type: string
  isArray: boolean
}

export interface TableInfo {
  name: string
  canInsert: boolean
  columns: readonly ColumnInfo[]
}

export default class SchemaInfo {
  constructor(
    private readonly pool: DatabasePoolType,
    public readonly name: string
  ) { }

  public tableQuery(): TaggedTemplateLiteralInvocationType<Omit<TableInfo, 'columns'>> {
    return sql<Omit<TableInfo, 'columns'>>`
      SELECT table_name AS "name"
          , (is_insertable_into = 'YES') AS "canInsert"
          , pg_catalog.obj_description(('"' || table_name || '"')::regclass::oid) AS "comment"
      FROM information_schema.tables
      WHERE table_schema = ${this.name}
      ORDER BY table_name ASC
    `
  }

  public columnQuery(table: string): TaggedTemplateLiteralInvocationType<ColumnInfo> {
    return sql<ColumnInfo>`
      SELECT c.column_name AS "name"
           , (c.is_nullable = 'YES') AS "nullable"
           , (c.column_default IS NOT NULL) AS "hasDefault"
           , c.ordinal_position AS "order"
           , COALESCE(de.udt_name, dc.udt_name, e.udt_name, e.data_type, c.udt_name, c.data_type) AS "type"
           , (c.data_type = 'ARRAY') AS "isArray"
           , pg_catalog.col_description(('"' || c.table_name || '"')::regclass::oid, c.ordinal_position::int) AS "comment"
      FROM
        information_schema.columns c LEFT JOIN
        information_schema.element_types e ON (
          c.data_type = 'ARRAY' AND
          c.table_schema = e.object_schema AND
          c.table_name = e.object_name AND
          c.dtd_identifier = e.collection_type_identifier
        ) LEFT JOIN
        information_schema.domains dc ON (
          c.data_type = 'USER-DEFINED' AND
          c.table_schema = dc.domain_schema AND
          c.domain_name = dc.domain_name
        ) LEFT JOIN
        information_schema.domains de ON (
          e.data_type = 'USER-DEFINED' AND
          e.object_schema = de.domain_schema AND
          e.udt_name = de.domain_name
        )
      WHERE (
        c.table_schema = ${this.name} AND
        c.table_name = ${table}
      )
      ORDER BY c.ordinal_position
    `
  }

  public enumQuery(): TaggedTemplateLiteralInvocationType<EnumInfo> {
    return sql<EnumInfo>`
      SELECT t.typname AS "name"
           , array_agg(e.enumlabel)::TEXT[] AS "values"
           , pg_catalog.obj_description(e.enumtypid) AS "comment"
      FROM
        pg_type t JOIN
        pg_enum e ON (
          t.oid = e.enumtypid
        ) JOIN
        pg_catalog.pg_namespace n ON (
          n.oid = t.typnamespace
        )
      WHERE (
        n.nspname = ${this.name}
      )
      GROUP BY n.nspname
             , t.typname
             , e.enumtypid
      ORDER BY t.typname
    `
  }

  public async disconnect(): Promise<void> {
    await this.pool.end()
  }

  public async getTables(): Promise<readonly TableInfo[]> {
    const tables = await this.pool.any(this.tableQuery())

    return Promise.all(
      tables.map<Promise<TableInfo>>(async (table) => {
        return {
          ...table,
          columns: await this.getTableColumns(table.name)
        }
      })
    )
  }

  public async getTableColumns(table: string): Promise<readonly ColumnInfo[]> {
    return await this.pool.many(this.columnQuery(table))
  }

  public async getEnums(): Promise<readonly EnumInfo[]> {
    return await this.pool.any(this.enumQuery())
  }

}
