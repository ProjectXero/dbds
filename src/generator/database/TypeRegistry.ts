import { factory, TypeNode } from 'typescript'

export const UNKNOWN = factory.createTypeReferenceNode('unknown')

export const defaultTypeMap = Object.freeze({
  bpchar: 'string',
  char: 'string',
  varchar: 'string',
  text: 'string',
  citext: 'string',
  uuid: 'string',
  bytea: 'string',
  inet: 'string',
  time: 'string',
  timetz: 'string',
  interval: 'string',
  name: 'string',
  int2: 'number',
  int4: 'number',
  int8: 'number',
  float4: 'number',
  float8: 'number',
  numeric: 'number',
  money: 'number',
  oid: 'number',
  bool: 'boolean',
  json: 'unknown',
  jsonb: 'unknown',
  date: 'Date',
  timestamp: 'Date',
  timestamptz: 'Date',
})

export default class TypeRegistry {
  public readonly typeMap: Record<string, string>

  constructor() {
    this.typeMap = { ...defaultTypeMap }
  }

  public has(typename: string): boolean {
    return Object.prototype.hasOwnProperty.call(this.typeMap, typename)
  }

  public get(typename: string): TypeNode {
    if (this.has(typename)) {
      return factory.createTypeReferenceNode(this.typeMap[typename])
    }

    console.warn(`Unknown type detected: '${typename}'. You have either disabled generation of the type or this is a bug.`)
    this.add(typename, 'unknown')

    return UNKNOWN
  }

  public add(typename: string, target: string): void {
    if (this.has(typename)) {
      console.warn(`Re-registering known type '${typename}': ${this.typeMap[typename]} => ${target}`)
    }

    this.typeMap[typename] = target;
  }
}
