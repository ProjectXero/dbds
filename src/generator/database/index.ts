export { default as SchemaInfo } from './SchemaInfo'
export * from './SchemaInfo'
export { default as TypeRegistry } from './TypeRegistry'
