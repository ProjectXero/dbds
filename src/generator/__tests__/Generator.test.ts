import { createMockPool } from '../../testing'

import { SchemaInfo } from '../database'
import { Generator } from '..'

import mockEnums from './__fixtures__/enums'
import mockTables from './__fixtures__/tables'

describe(Generator, () => {
  const dummySchema = new SchemaInfo(createMockPool(), 'any')
  const tableSpy = jest.spyOn(dummySchema, 'getTables')
  const enumSpy = jest.spyOn(dummySchema, 'getEnums')

  const warnSpy = jest.spyOn(console, 'warn')
  warnSpy.mockImplementation(() => {
    /* do nothing */
  })

  const errorSpy = jest.spyOn(console, 'error')
  errorSpy.mockImplementation(() => {
    /* do nothing */
  })

  beforeEach(() => {
    warnSpy.mockClear()
    tableSpy.mockImplementation(async () => mockTables)
    enumSpy.mockImplementation(async () => mockEnums)
  })

  describe('with default options', () => {
    const instance = new Generator({
      schema: dummySchema,
    })

    it('generates metadata and select/insert schemas', async () => {
      expect(await instance.build()).toMatchSnapshot()
    })
  })

  describe('with everything disabled', () => {
    const instance = new Generator({
      schema: dummySchema,
      genTableMetadata: false,
      genSelectSchemas: false,
      genInsertSchemas: false,
      genUpdateSchemas: false,
      genEnums: false,
      genInsertTypes: false,
      genTables: false,
      genTypeObjects: false,
      genSchemaObjects: false,
    })

    it('generates nothing', async () => {
      expect(await instance.build()).toMatchSnapshot()
    })
  })

  describe('with tables enabled', () => {
    const instance = new Generator({
      schema: dummySchema,
      genTableMetadata: false,
      genSelectSchemas: false,
      genInsertSchemas: false,
      genUpdateSchemas: false,
      genEnums: false,
      genInsertTypes: false,
      genTables: true,
      genTypeObjects: false,
      genSchemaObjects: false,
    })

    it('generates tables', async () => {
      expect(await instance.build()).toMatchSnapshot()
      expect(warnSpy.mock.calls).toMatchSnapshot()
    })
  })

  describe('with enums enabled', () => {
    const instance = new Generator({
      schema: dummySchema,
      genTableMetadata: false,
      genSelectSchemas: false,
      genInsertSchemas: false,
      genUpdateSchemas: false,
      genEnums: true,
      genInsertTypes: false,
      genTables: false,
      genTypeObjects: false,
      genSchemaObjects: false,
    })

    it('generates enums', async () => {
      expect(await instance.build()).toMatchSnapshot()
    })
  })

  describe('with insert types enabled', () => {
    const instance = new Generator({
      schema: dummySchema,
      genTableMetadata: false,
      genSelectSchemas: false,
      genInsertSchemas: false,
      genUpdateSchemas: false,
      genEnums: false,
      genInsertTypes: true,
      genTables: false,
      genTypeObjects: false,
      genSchemaObjects: false,
    })

    it('generates insert types', async () => {
      expect(await instance.build()).toMatchSnapshot()
      expect(warnSpy.mock.calls).toMatchSnapshot()
    })
  })

  describe('with type objects enabled', () => {
    const instance = new Generator({
      schema: dummySchema,
      genTableMetadata: false,
      genSelectSchemas: false,
      genInsertSchemas: false,
      genUpdateSchemas: false,
      genEnums: false,
      genInsertTypes: false,
      genTables: false,
      genTypeObjects: true,
      genSchemaObjects: false,
    })

    it('generates type objects', async () => {
      expect(await instance.build()).toMatchSnapshot()
      expect(warnSpy.mock.calls).toMatchSnapshot()
    })
  })

  describe('with schema objects enabled', () => {
    const instance = new Generator({
      schema: dummySchema,
      genTableMetadata: false,
      genSelectSchemas: false,
      genInsertSchemas: false,
      genUpdateSchemas: false,
      genEnums: false,
      genInsertTypes: false,
      genTables: false,
      genTypeObjects: false,
      genSchemaObjects: true,
    })

    it('generates schema objects', async () => {
      expect(await instance.build()).toMatchSnapshot()
      expect(warnSpy.mock.calls).toMatchSnapshot()
    })
  })

  describe('with different case conversions', () => {
    const instance = new Generator({
      schema: dummySchema,
      transformColumns: 'camel',
      transformEnumMembers: 'none',
    })

    it('properly cases members', async () => {
      expect(await instance.build()).toMatchSnapshot()
    })
  })

  describe('when re-registering an already-registered type', () => {
    beforeEach(() => {
      tableSpy.mockImplementation(async () => [
        {
          name: 'somehow_duplicated_type_name',
          canInsert: true,
          columns: [],
        },
        {
          name: 'somehow_duplicated_type_name',
          canInsert: true,
          columns: [],
        },
      ])
    })

    const instance = new Generator({
      schema: dummySchema,
      genEnums: false,
      genTables: true,
      genInsertTypes: false,
      genTypeObjects: false,
    })

    it('outputs a warning', async () => {
      expect(await instance.build()).toMatchSnapshot()
      expect(warnSpy.mock.calls).toMatchSnapshot()
    })
  })

  describe('build', () => {
    const instance = new Generator({
      schema: dummySchema,
    })
    const tableBuildersSpy = jest.spyOn(instance, 'tableBuilders' as never)

    beforeEach(() => {
      tableBuildersSpy.mockImplementation(() => {
        throw new Error('testing error handling')
      })
    })

    afterEach(() => {
      tableBuildersSpy.mockReset()
    })

    describe('when builders throw an error', () => {
      it('logs the error and rethrows', async () => {
        await expect(instance.build()).rejects.toThrowError(
          'testing error handling'
        )
        expect(errorSpy.mock.calls).toMatchSnapshot()
      })
    })
  })
})
