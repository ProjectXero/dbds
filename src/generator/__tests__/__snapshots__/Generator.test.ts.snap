// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator with default options generates everything 1`] = `
"export type PrimitiveValueType = string | number | boolean | null;
export type SimpleValueType = PrimitiveValueType | Date;
export type SerializableValueType = SimpleValueType | {
    [key in string]: SerializableValueType | undefined;
} | Array<SerializableValueType> | ReadonlyArray<SerializableValueType>;
export enum TestEnum {
    A = \\"A\\",
    B = \\"b\\",
    CamelCaseRules = \\"camel_case_rules\\"
}
export interface TableWithNoColumns {
}
export interface TableWithNoColumns$Insert {
}
export const TableWithNoColumns$Types: Record<keyof TableWithNoColumns, string> = {};
export interface TableWithNumericId {
    id: number;
}
export interface TableWithNumericId$Insert {
    id?: number;
}
export const TableWithNumericId$Types: Record<keyof TableWithNumericId, string> = {
    id: \\"int8\\"
};
export interface TableWithCustomTypes {
    enum_type: TestEnum;
    enum_array_type: TestEnum[];
    table_type: TableWithUuidId;
    table_array_type: TableWithUuidId[];
}
export interface TableWithCustomTypes$Insert {
    enum_type: TestEnum;
    enum_array_type: TestEnum[];
    table_type: TableWithUuidId;
    table_array_type: TableWithUuidId[];
}
export const TableWithCustomTypes$Types: Record<keyof TableWithCustomTypes, string> = {
    enum_type: \\"test_enum\\",
    enum_array_type: \\"test_enum\\",
    table_type: \\"table_with_uuid_id\\",
    table_array_type: \\"table_with_uuid_id\\"
};
export interface TableWithUuidId {
    id: string;
}
export interface TableWithUuidId$Insert {
    id?: string;
}
export const TableWithUuidId$Types: Record<keyof TableWithUuidId, string> = {
    id: \\"uuid\\"
};
export interface TableWithNullableFields {
    nullable: string | null;
    nullable_with_default: string | null;
    nullable_array: string[] | null;
    nullable_array_with_default: string[] | null;
}
export interface TableWithNullableFields$Insert {
    nullable?: string | null;
    nullable_with_default?: string | null;
    nullable_array?: string[] | null;
    nullable_array_with_default?: string[] | null;
}
export const TableWithNullableFields$Types: Record<keyof TableWithNullableFields, string> = {
    nullable: \\"text\\",
    nullable_with_default: \\"text\\",
    nullable_array: \\"text\\",
    nullable_array_with_default: \\"text\\"
};
"
`;

exports[`Generator with different case conversions properly cases members 1`] = `
"export type PrimitiveValueType = string | number | boolean | null;
export type SimpleValueType = PrimitiveValueType | Date;
export type SerializableValueType = SimpleValueType | {
    [key in string]: SerializableValueType | undefined;
} | Array<SerializableValueType> | ReadonlyArray<SerializableValueType>;
export enum TestEnum {
    A = \\"A\\",
    B = \\"b\\",
    CAMEL_CASE_RULES = \\"camel_case_rules\\"
}
export interface TableWithNoColumns {
}
export interface TableWithNoColumns$Insert {
}
export const TableWithNoColumns$Types: Record<keyof TableWithNoColumns, string> = {};
export interface TableWithNumericId {
    id: number;
}
export interface TableWithNumericId$Insert {
    id?: number;
}
export const TableWithNumericId$Types: Record<keyof TableWithNumericId, string> = {
    id: \\"int8\\"
};
export interface TableWithCustomTypes {
    enumType: TestEnum;
    enumArrayType: TestEnum[];
    tableType: TableWithUuidId;
    tableArrayType: TableWithUuidId[];
}
export interface TableWithCustomTypes$Insert {
    enumType: TestEnum;
    enumArrayType: TestEnum[];
    tableType: TableWithUuidId;
    tableArrayType: TableWithUuidId[];
}
export const TableWithCustomTypes$Types: Record<keyof TableWithCustomTypes, string> = {
    enumType: \\"test_enum\\",
    enumArrayType: \\"test_enum\\",
    tableType: \\"table_with_uuid_id\\",
    tableArrayType: \\"table_with_uuid_id\\"
};
export interface TableWithUuidId {
    id: string;
}
export interface TableWithUuidId$Insert {
    id?: string;
}
export const TableWithUuidId$Types: Record<keyof TableWithUuidId, string> = {
    id: \\"uuid\\"
};
export interface TableWithNullableFields {
    nullable: string | null;
    nullableWithDefault: string | null;
    nullableArray: string[] | null;
    nullableArrayWithDefault: string[] | null;
}
export interface TableWithNullableFields$Insert {
    nullable?: string | null;
    nullableWithDefault?: string | null;
    nullableArray?: string[] | null;
    nullableArrayWithDefault?: string[] | null;
}
export const TableWithNullableFields$Types: Record<keyof TableWithNullableFields, string> = {
    nullable: \\"text\\",
    nullableWithDefault: \\"text\\",
    nullableArray: \\"text\\",
    nullableArrayWithDefault: \\"text\\"
};
"
`;

exports[`Generator with enums enabled generates enums 1`] = `
"export type PrimitiveValueType = string | number | boolean | null;
export type SimpleValueType = PrimitiveValueType | Date;
export type SerializableValueType = SimpleValueType | {
    [key in string]: SerializableValueType | undefined;
} | Array<SerializableValueType> | ReadonlyArray<SerializableValueType>;
export enum TestEnum {
    A = \\"A\\",
    B = \\"b\\",
    CamelCaseRules = \\"camel_case_rules\\"
}
"
`;

exports[`Generator with everything disabled generates nothing 1`] = `
"export type PrimitiveValueType = string | number | boolean | null;
export type SimpleValueType = PrimitiveValueType | Date;
export type SerializableValueType = SimpleValueType | {
    [key in string]: SerializableValueType | undefined;
} | Array<SerializableValueType> | ReadonlyArray<SerializableValueType>;
"
`;

exports[`Generator with insert types enabled generates insert types 1`] = `
"export type PrimitiveValueType = string | number | boolean | null;
export type SimpleValueType = PrimitiveValueType | Date;
export type SerializableValueType = SimpleValueType | {
    [key in string]: SerializableValueType | undefined;
} | Array<SerializableValueType> | ReadonlyArray<SerializableValueType>;
export interface TableWithNoColumns$Insert {
}
export interface TableWithNumericId$Insert {
    id?: number;
}
export interface TableWithCustomTypes$Insert {
    enum_type: unknown;
    enum_array_type: unknown[];
    table_type: unknown;
    table_array_type: unknown[];
}
export interface TableWithUuidId$Insert {
    id?: string;
}
export interface TableWithNullableFields$Insert {
    nullable?: string | null;
    nullable_with_default?: string | null;
    nullable_array?: string[] | null;
    nullable_array_with_default?: string[] | null;
}
"
`;

exports[`Generator with insert types enabled generates insert types 2`] = `
Array [
  Array [
    "Unknown type detected: 'test_enum'. You have either disabled generation of the type or this is a bug.",
  ],
  Array [
    "Unknown type detected: 'test_enum'. You have either disabled generation of the type or this is a bug.",
  ],
  Array [
    "Unknown type detected: 'table_with_uuid_id'. You have either disabled generation of the type or this is a bug.",
  ],
]
`;

exports[`Generator with tables enabled generates tables 1`] = `
"export type PrimitiveValueType = string | number | boolean | null;
export type SimpleValueType = PrimitiveValueType | Date;
export type SerializableValueType = SimpleValueType | {
    [key in string]: SerializableValueType | undefined;
} | Array<SerializableValueType> | ReadonlyArray<SerializableValueType>;
export interface TableWithNoColumns {
}
export interface TableWithNumericId {
    id: number;
}
export interface TableWithCustomTypes {
    enum_type: unknown;
    enum_array_type: unknown[];
    table_type: TableWithUuidId;
    table_array_type: TableWithUuidId[];
}
export interface TableWithUuidId {
    id: string;
}
export interface TableWithNullableFields {
    nullable: string | null;
    nullable_with_default: string | null;
    nullable_array: string[] | null;
    nullable_array_with_default: string[] | null;
}
"
`;

exports[`Generator with tables enabled generates tables 2`] = `
Array [
  Array [
    "Unknown type detected: 'test_enum'. You have either disabled generation of the type or this is a bug.",
  ],
]
`;

exports[`Generator with type objects enabled generates type objects 1`] = `
"export type PrimitiveValueType = string | number | boolean | null;
export type SimpleValueType = PrimitiveValueType | Date;
export type SerializableValueType = SimpleValueType | {
    [key in string]: SerializableValueType | undefined;
} | Array<SerializableValueType> | ReadonlyArray<SerializableValueType>;
export const TableWithNoColumns$Types: Record<keyof TableWithNoColumns, string> = {};
export const TableWithNumericId$Types: Record<keyof TableWithNumericId, string> = {
    id: \\"int8\\"
};
export const TableWithCustomTypes$Types: Record<keyof TableWithCustomTypes, string> = {
    enum_type: \\"test_enum\\",
    enum_array_type: \\"test_enum\\",
    table_type: \\"table_with_uuid_id\\",
    table_array_type: \\"table_with_uuid_id\\"
};
export const TableWithUuidId$Types: Record<keyof TableWithUuidId, string> = {
    id: \\"uuid\\"
};
export const TableWithNullableFields$Types: Record<keyof TableWithNullableFields, string> = {
    nullable: \\"text\\",
    nullable_with_default: \\"text\\",
    nullable_array: \\"text\\",
    nullable_array_with_default: \\"text\\"
};
"
`;

exports[`Generator with type objects enabled generates type objects 2`] = `
Array [
  Array [
    "Unknown type detected: 'test_enum'. You have either disabled generation of the type or this is a bug.",
  ],
  Array [
    "Unknown type detected: 'test_enum'. You have either disabled generation of the type or this is a bug.",
  ],
  Array [
    "Unknown type detected: 'table_with_uuid_id'. You have either disabled generation of the type or this is a bug.",
  ],
]
`;
